### PR TITLE
Preparation for m-c merge (fixes #36, fixes #38, fixes #39)

### LINKED ISSUE #36: Move js/src/frontend-rs to js/src/frontend/smoosh
2 issues:
  * https://github.com/mozilla-spidermonkey/jsparagus/issues/247
  * having `frontend` and `frontend-rs` in the same directory isn't good for autocomplete

### LINKED ISSUE #38: Rename jsparagus and rust-frontend to smoosh
As a preparation to add configure option

### LINKED ISSUE #39: Add configure option for smoosh

### PR DESCRIPTION
**Pull request by arai-a**
This patch stack includes https://github.com/mozilla-spidermonkey/rust-frontend/pull/35 that is completely unrelated.
will rebase after it gets merged.

the remaining parts are:
 * directory renaming
 * class/option renaming
 * configure option and conditional build

jsparagus submodule is still kept, but now just moving it outside (and remove submodule) should work.

### CODE REVIEW

File: js/moz.configure [hunk 1]
```diff
@@ -601,3 +601,16 @@ def enable_new_regexp(value):
 
 set_config('ENABLE_NEW_REGEXP', enable_new_regexp)
 set_define('ENABLE_NEW_REGEXP', enable_new_regexp)
+
+# SmooshMonkey (new frontend)
+# ==================================================
+
+js_option('--enable-smoosh', default=False, help='Enable SmooshMonkey')
```
Comment (jorendorff) [inline js/moz.configure]: `help='Enable SmooshMonkey (new JS engine frontend)'`

File: js/src/shell/js.cpp [hunk 2]
```diff
@@ -11476,10 +11486,14 @@ int main(int argc, char** argv, char** envp) {
                                "Dynamically load LIBRARY") ||
       !op.addBoolOption('\0', "suppress-minidump",
                         "Suppress crash minidumps") ||
+#ifdef JS_ENABLE_SMOOSH
+      !op.addBoolOption('\0', "smoosh", "Use SmooshMonkey") ||
```
Comment (jorendorff) [inline js/src/shell/js.cpp]: `"Use SmooshMonkey frontend"`

File: js/src/frontend/smoosh/smoosh.h [hunk 3]
```diff
@@ -3,9 +3,9 @@
 #include <cstdlib>
 #include <new>
 
-template <typename T>
+template<typename T>
 struct CVec {
-  T* data;
+  T *data;
```
Comment (jorendorff) [inline js/src/frontend/smoosh/smoosh.h]: It looks like this generated file wasn't clang-formatted on commit. I don't know if it should be or not, but it probably shouldn't flip-flop.

**Review by jorendorff — APPROVED** [review]
OK, please add a commit that updates jsparagus and land this. The review comments attached here are optional (and even if you do want to include them, they could land in a follow-up PR).

File: js/src/frontend/smoosh/smoosh.h (hunk 3 above)
Comment (arai-a) [reply js/src/frontend/smoosh/smoosh.h]: I'll add this to https://searchfox.org/mozilla-central/source/.clang-format-ignore

**Pull request merged**